Rethink UI and core design

saagarjha opened the pull request:
Please do not merge this; I'm not done with it. This was just to let you check out some of the changes I think we need:

* Improve the main window's UI
* Centralize the settings and share it between the app, extension, and CLI
* Improve parser code
* Reduce API surface area

Jintin commented:
It would be better if you split into several pull request.

saagarjha commented:
@Jintin I apologize for the long period of inactivity–I'm still working on this and I was just sidetracked by other stuff. Would you mind reopening this pull request?

Jintin commented:
Sure. or you can create a another one.

saagarjha commented:
@Jintin I'm finding it difficult to split this pull request apart into parts that can compile. If you'd like, I can:

1. Submit everything as one pull request, as it is now (once I finish, of course). I can split it into multiple commits, though.
2. Split it into multiple pull requests which will not compile by themselves.

Which do you prefer?

saagarjha commented:
@Jintin I think I'm done with this. There's a lot in it:

* New UI inspired by the old one. There's now a tab view with installation instructions and preferences, and a preview in a new split view window that allows users to input their own code. 
* Centralized preferences that are shared between the app extension and command line tool. 
* New testing format, which I hope is more extensible. It dynamically adds tests based on file-based test cases (I've converted your existing test cases to this format; please give them appropriate numbers as well as add a few "simple" ones as I did, such as a structure, function, for loop, etc.)
* Minor bug fixes and API enhancements

Please ignore the build warnings, I have a Swift 4 pull request ready that fixes these once this is merged.

saagarjha commented:
Screenshots of the new UI:
![screen shot 2017-07-06 at 16 33 57](https://user-images.githubusercontent.com/13786931/27937091-90d3f668-6269-11e7-8a92-46c509e26bfc.png)
![screen shot 2017-07-06 at 16 34 00](https://user-images.githubusercontent.com/13786931/27937090-90d28f44-6269-11e7-8bd0-5dcd6118e0b4.png)
<img width="912" alt="screen shot 2017-07-06 at 16 38 12" src="https://user-images.githubusercontent.com/13786931/27937092-90d8632e-6269-11e7-8636-375e0687436e.png">

Jintin commented:
@saagarjha Please give me couple days to review.
First of all, please format and use 4 space to indent.
Thank you for your patience and great work.

saagarjha commented:
> First of all, please format and use 4 space to indent.

One would have thought that I would have remembered this when working on a *formatter*…

> Please give me couple days to review.

Sure, no thing. If you need clarification on anything please let me know.

Jintin commented on Swimat/SwimatViewController.swift:
```diff
@@ -72,30 +64,41 @@ class SwimatViewController: NSViewController {
     func refreshInstallButton() {
         // Check for swimat, fileExists(atPath:) returns false for symlinks
-        if (try? FileManager.default.attributesOfItem(atPath: "\(installPath)swimat")) != nil {
+        if ((try? FileManager.default.attributesOfItem(atPath: "\(installPath)swimat")) != nil) {
```
no need to add `(` and `)`

saagarjha replied:
Will remove

Jintin commented on Parser/Preferences.swift:
```diff
@@ -0,0 +1,32 @@
+import Foundation
+
+class Preferences: Codable {
```
Currently we still on Swift 3.0 XD

saagarjha replied:
Will polyfill

Jintin commented on Parser/Indent.swift:
```diff
@@ -19,9 +19,13 @@ enum IndentType: Character {
+    }
     static var size: Int = 0
-    static var paraAlign = true
+    static var paraAlign = false
```
Maybe we can use `true` as default since apple does too.

saagarjha replied:
I think that settings should be off by default, in general. If someone wants Swimat to do something for them they should turn it on. This follows the principle of least surprise.

Jintin replied:
OK

Jintin commented on Extension/SourceEditorCommand.swift:
```diff
@@ -43,7 +40,7 @@ class SourceEditorCommand: NSObject, XCSourceEditorCommand {
                 if let selection = selections[i] as? XCSourceTextRange, selection.start != selection.end {
                     hasSelection = true
-                    for j in selection.start.line ... selection.end.line {
+                    for j in selection.start.line...selection.end.line {
```
space around  `...` ?

saagarjha replied:
Isn't that how it's supposed to be? At least, that's the style I normally see.

Jintin replied:
I mean why you remove the space around it.

saagarjha replied:
This is how [Apple does it](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/BasicOperators.html#//apple_ref/doc/uid/TP40014097-CH6-ID73).

saagarjha commented:
@Jintin Resolved merge conflict.

saagarjha commented:
@Jintin Is there anything you'd like me to change?

Jintin commented on Extension/Supporting Files/Extension.entitlements:
```diff
@@ -6,7 +6,7 @@
 	<key>com.apple.security.application-groups</key>
 	<array>
-		<string>com.jintin.swimat.config</string>
+		<string>com.jintin.swimat.configuration</string>
```
please keep the old setting

saagarjha replied:
Umm, ok…

Jintin commented:
Could you separate your commit in several pull request like: test, layout, pref... etc.
And since you use swift 4 syntax, may be I will merge in another branch.

saagarjha commented:
> Could you separate your commit in several pull request like: test, layout, pref... etc.

I could, but the commits won't compile by themselves. Only the final one will. Is that okay?

> And since you use swift 4 syntax, may be I will merge in another branch.

Actually this is still Swift 3…I have a Swift 4 version ready locally, but I haven't pushed it yet.

EDIT: Oh, you're talking about `Codable`. I'll put a polyfill in for it.

Jintin commented:
I encourage make every commit can compile is possible.
We should make the highest standard quality of pull request as we are the main contributors.
And for separate pull requests you can focus on the swift 3 compatible parts.

Jintin commented on Extension/SourceEditorCommand.swift:
```diff
@@ -4,24 +4,21 @@ import XcodeKit
 
         let uti = invocation.buffer.contentUTI
-        if uti != "com.apple.dt.playground" && uti != "public.swift-source" && uti != "com.apple.dt.playgroundpage"{
+        guard uti == "com.apple.dt.playground" || uti == "public.swift-source" || uti == "com.apple.dt.playgroundpage" else {
```
There's a bug before.
Maybe multi `guard let` is too complicate and not best choice in this case.
Array contain maybe another choice.

saagarjha replied:
Hmm, what bug are you talking about? In any case the UTI checking code needs to be rewritten; I'll do this after this+a couple other pull requests are merged.

Jintin replied:
There is no bug now definitely.
What I mean is that multi `guard let` is too complicate to understand.
```swift
guard let utiList.contain(uti) else {
    return
}
```

saagarjha replied:
I see your point. I'll leave it for now, since I'll write a better, more correct solution in a bit.

saagarjha commented:
@Jintin I'll try my best, but it will take a while.

saagarjha committed:
General improvements

Fix code formatting, remove unnecessary resources, and improve code fluency

Jintin commented:
You can take anytime you need.
Thank you for your help.

saagarjha committed:
Improve preferences management

Share preferences across all targets, improve API consistency, and allow for
"testing" configurations that are separate from the user's preferences

saagarjha committed:
Modernize UI

Improve the UI with visual cues taken from earlier renditions

saagarjha committed:
Modularize testing

Tests are now dynamically generated from a set of files instead of hard-coded
into the codebase

saagarjha commented:
@Jintin I've split the changes into separate commits (and learned a lot about `git add -p`!). Please take a look at it.

Jintin commented:
Good for you. But I mean separate in different pull request by functionality.
So we can merge in different stage (before swift 4 and after ... etc).

saagarjha commented:
OK, but you'll have to merge them in order. Also, just FYI, these should all compile in Swift 3.2 as well.

saagarjha commented:
@Jintin #169, #170, #171, #172, #173, and #174.

d2burke commented:
Just read through this thread. Nothing exciting here, just two devs taking on a cool project and working through conflict and compromise together. Love it. 

Just wanted to say I appreciated the exemplary collaboration.
